sklearn/gaussian_process/gpr.py

```diff
@@ -0,0 +1,171 @@
+
+        return self
+
+    def predict(self, X, return_std=False, return_cov=False):
+        if return_std and return_cov:
+            raise RuntimeError(
+                "Not returning standard deviation of predictions when "
+                "returning full covariance.")
+
+        X = check_array(X)
+
+        if not hasattr(self, "X_fit_"):  # Unfitted; predict based on GP prior
+            y_mean = np.zeros(X.shape[0])
+            if return_cov:
+                y_cov = self.kernel(X)
+                return y_mean, y_cov
+            elif return_std:
+                # XXX: Compute y_std more efficiently
+                y_std = np.sqrt(np.diag(self.kernel(X)))
+                return y_mean, y_std
+            else:
+                return y_mean
+        else:  # Predict based on GP posterior
+            K_trans = self.kernel(X, self.X_fit_)
+            y_mean = K_trans.dot(self.alpha_)  # Line 4 (y_mean = f_star)
+            if return_cov:
+                v = cho_solve((self.L_, True), K_trans.T)  # Line 5
+                y_cov = \
+                    self.kernel(X) - K_trans.dot(v)  # Line 6
+                return y_mean, y_cov
+            elif return_std:
+                # XXX: Compute y_std more efficiently
+                v = cho_solve((self.L_, True), K_trans.T)  # Line 5
+                y_cov = \
+                    self.kernel(X) - K_trans.dot(v)  # Line 6
+                y_std = np.sqrt(np.diag(y_cov))
+                return y_mean, y_std
+            else:
+                return y_mean
+
+    def sample(self, X, n_samples=1, random_state=0):
+        rng = check_random_state(random_state)
+
+        y_mean, y_cov = self.predict(X, return_cov=True)
+        y_samples = rng.multivariate_normal(y_mean, y_cov, n_samples).T
+        return y_samples
+
+    def log_marginal_likelihood(self, theta, eval_gradient=False):
+        import copy  # XXX: Avoid deepcopy
+        kernel = copy.deepcopy(self.kernel)
+        kernel.params = theta
+
+        if eval_gradient:
+            K, K_gradient = kernel(self.X_fit_, eval_gradient=True)
+        else:
+            K = kernel(self.X_fit_)
+
+        K[np.diag_indices_from(K)] += self.y_err
+        try:
+            L = cholesky(K, lower=True)  # Line 2
+        except np.linalg.LinAlgError:
+            return (-np.inf, np.zeros_like(theta))\
+                 if eval_gradient else -np.inf
+
+        alpha = cho_solve((L, True), self.y_fit_)  # Line 3
+
+        # Compute log-likelihood (compare line 7)
+        log_likelihood = -0.5*self.y_fit_.dot(alpha)
+        log_likelihood -= np.log(np.diag(L)).sum()
+        log_likelihood -= K.shape[0] / 2 * np.log(2 * np.pi)
+
+        if eval_gradient:  # compare Equation 5.9 from GPML
+            tmp = np.outer(alpha, alpha)
+            tmp -= cho_solve((L, True), np.eye(K.shape[0]))
+            gradient = 0.5 * np.einsum("ij,ijk->k", tmp, K_gradient)
```

Comment:
Depending on the size of the problem and the ordering of the arrays, you may gain a little more with `tmp.ravel().dot(K_gradient.reshape(-1, K_gradient.shape[2]))`. In order to ensure that the array is not copied, you can also do

```python
tmp_ = tmp.view()
tmp_.shape = -1,
K_gradient_ = K_gradient.view()
K_gradient.shape = (-1, K_gradient.shape[2])

gradient = 0.5 * tmp_dot(K_gradient_)
```